### DISCUSSION TITLE
Find relation between beam and wall

### DISCUSSION THREAD

**Discussion by sanchezzuluaga** [IfcOpenShell/IfcOpenShell#3196] [Q&A] [answered]
Hi everyone,

I'm currently working on identifying the support walls for a beam in a given structure. I'm wondering if there's a Python-based solution that can determine the supporting walls without requiring coordinates and space information.

Any insights or suggestions would be greatly appreciated!

Thank you!

**Comment by aothms** [comment]
My guess would be that this really depends on the modeller and modelling application. While there might be semantic relationships in the ifc schema to encode this kind of information, it would really depend on the modeller also modelling this as such and the tool being able to export that.

With `ifcopenshell.geom.tree` exists a reasonably straightforward way of querying this from the geometric perspective.

**Comment by sanchezzuluaga** [comment]
Hi Thomas,

Thank you for your prompt response. I have been exploring the implementation of ifcopenshell.geom.tree, but unfortunately, I haven't found comprehensive documentation or examples to assist me. However, I did discover some useful resources on processing the geometry of building elements. Considering this information, I suggest creating custom functions to extract the coordinates of walls and beams. This approach will enable to check for intersections and determine if the beam is supported by specific walls.

If you have any insights or suggestions regarding the use of ifcopenshell. geom.tree, I would greatly appreciate them. Additionally, I would highly appreciate your guidance if you are aware of a more straightforward solution for the task at hand.

Thank you!

Best regards,
Simon

**Reply by aothms** [reply]
https://blenderbim.org/docs-python/autoapi/ifcopenshell/geom/main/index.html#ifcopenshell.geom.main.tree

https://community.osarch.org/discussion/733/relationships-in-ifc

Any help improving the docs is always welcome :)

**Reply by Moult** [reply]
@sanchezzuluaga would you be interested in writing some docs about how the geom tree works?

**Reply by sanchezzuluaga** [reply]
HI @Moult , sure! Once I understand how it works I can do that.

**Reply by sanchezzuluaga** [reply]
Hello everyone,

I attempted to generate the tree structure for a wall following the instructions provided in this discussion: [link](https://community.osarch.org/discussion/733/relationships-in-ifc).  However, I encountered an issue where I obtained an empty list instead of the expected tree structure.

To provide some context, I am using version 0.7.0.230523. of IfcOpenShell.

Here is an example of the code I used:

```ruby
walls = file.by_type("IfcWall")
wall = walls[0]
matrix = Placement.get_local_placement(wall.ObjectPlacement)

location = tuple(map(float, matrix[0:3,3]))
print(location)

# setup BVH tree
tree_settings = Geometry.settings()
#print(tree_settings)

tree_settings.set(tree_settings.DISABLE_TRIANGULATION, True)
tree_settings.set(tree_settings.DISABLE_OPENING_SUBTRACTIONS, True)
it = Geometry.iterator(tree_settings, file, include=("IfcSpace"))
print(it)

geometric_tree = Geometry.main.tree(file=file, settings=tree_settings)


t = Geometry.tree() 
t.add_iterator(it)

# # # search tree
a = t.select(location, extend=0.4)

print(a)

```

**Reply by aothms** [reply]
One potential issue might be units, see `calculate_unit_scale()`  usage in https://community.osarch.org/discussion/733/relationships-in-ifc (but it's not necessarily *the* issue).

**Reply by sanchezzuluaga** [reply]
Thank you for the prompt response. I have taken your suggestion and incorporated the ```calculate_unit_scale()```, but unfortunately, the resulting list remains empty.  I will provide an update here if I manage to find a solution.

**Reply by aothms** [reply]
If you provide the model I will also have a look.

**Reply by sanchezzuluaga** [reply]
Here is the model: 
[Birmer_2.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/11655021/Birmer_2.zip)

Thank you!

**Reply by aothms** [reply, marked as answer]
There's two issues:

- `assert file.by_type("IfcSpace")`, in the model you sent there is actually no IfcSpace elements so the tree remains empty
- `it = Geometry.iterator(tree_settings, file, include=("IfcSpace",))` On of the most annoying thing in the Python syntax. You cannot create a 1-size tuple without adding a comma.

```
>>> (1) == (1,)
False
>>> (1,2) == (1,2,)
True
```

**Reply by sanchezzuluaga** [reply]
Thank you for your answer! It works now

**Reply by Moult** [reply]
Awesome! If you have some spare time it would be super cool if you could help write a little bit in the docs about how to use the geometry tree or provide some sample code :)

**Reply by sanchezzuluaga** [reply]
Of course! I  will come back to the chat if I have further questions

**Comment by sanchezzuluaga** [comment]
Thank you for the information!
